src/Datasource/EntityTrait.php

```diff
@@ -242,13 +243,39 @@ public function set($property, $value = null, array $options = [])
         $options += ['setter' => true, 'guard' => $guard];
 
         foreach ($property as $p => $value) {
+            $nested = false;
+            if (strpos($p, '.') > 0 && !array_key_exists($p, $this->_properties)) {
+                list($p, $nested) = explode('.', $p, 2);
+            }
+
             if ($options['guard'] === true && !$this->accessible($p)) {
                 continue;
             }
 
             $this->dirty($p, true);
 
-            if (!array_key_exists($p, $this->_original) &&
+            if ($nested) {
+                if (!array_key_exists($p, $this->_properties)
+                    || is_null($this->_properties[$p])
+                ) {
+                    $value = new \Cake\ORM\Entity([$nested => $value]);
```

Comment:
Agreed, that's what it's looking like. It'll _have_ to throw a new exception since there's no other logical outcome if you ask for a nested path that descends into something that doesn't implement `getOriginal()` (EntityInterface) itself. That will also make it easier to switch back to ArrayObjects instead of Entities I think. Stand by please...